### ISSUE TITLE
Rector adds attributes which are already there or conflicting with what's already there

### ISSUE THREAD

**Issue by dkarlovi** [rectorphp/rector#8078]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | e.g. v0.17.6 (invoke `vendor/bin/rector --version`)             |

Because of https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6795 my PHPUnit tests need to still have annotations even though I'm actually using attributes. This doesn't bother PHPUnit, it works as expected.

But, Rector will keep add attributes, some of which are already there:

## Minimal PHP Code Causing Issue

Before rector:

```php
/**
 * @internal
 *
 * @coversNothing
 *
 * @small
 */
#[\PHPUnit\Framework\Attributes\CoversClass(ExtractorFactoryFactory::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(\App\Entity\User::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(UserIdentity::class)]
#[\PHPUnit\Framework\Attributes\Small]
final class ExtractorFactoryFactoryTest extends TestCase
```

After Rector:

```php
/**
 * @internal
 *
 *
 */
#[\PHPUnit\Framework\Attributes\CoversClass(ExtractorFactoryFactory::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(\App\Entity\User::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(UserIdentity::class)]
#[\PHPUnit\Framework\Attributes\Small]
#[\PHPUnit\Framework\Attributes\CoversNothing]
#[\PHPUnit\Framework\Attributes\Small]
final class ExtractorFactoryFactoryTest extends TestCase
```

## Expected Behaviour

1. don't add attributes which are already there (like `Small`)
2. don't add attributes which semantically conflict with attributes which are already there, in this case `CoversNothing` was blindly copied from the annotation, but it conflicts the `CoversClass` which is already there, it shouldn't be added at all.

```php
/**
 * @internal
 *
 *
 */
#[\PHPUnit\Framework\Attributes\CoversClass(ExtractorFactoryFactory::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(\App\Entity\User::class)]
#[\PHPUnit\Framework\Attributes\UsesClass(UserIdentity::class)]
#[\PHPUnit\Framework\Attributes\Small]
final class ExtractorFactoryFactoryTest extends TestCase
```

**Comment by TomasVotruba** [comment]
This is expected, as one of sets converts existing annotation to attribute. Problem is duplicate of annotation and attribute. Only one should be used.

You can also disable the responsble Rector set - PHPUnit 10 or use a [skip() option](https://getrector.com/documentation/ignoring-rules-or-paths) to convert annotations to attributes in tests. But I think using just one way of markers is more clear :)

**Comment by dkarlovi** [comment]
The **issue is not Rector removing the annotation**, the issue is Rector not checking what attributes it's adding.

In this case, it's just blindly converting the annotations to attributes, but the bug here **it shouldn't do that**. It should check what attributes already exist before adding new ones.

Basically:
1. Rector finds and removes annotation (because a rule said so), this is fine
2. it attempts to create the attribute (because the first step triggered this), this is fine
3. **(this is where the bug is)**
    1. currently, it just does it, without any regard for what the current attributes are, just appends the attribute, **this is the bug** reported here
    2. it should make checks to confirm the newly added attribute doesn't in some way conflict with attributes already defined on the class, in this case it would skip adding both of these attributes because the previous attributes prevented it

@TomasVotruba otherwise you're saying it's expected for Rector to duplicate attributes?

**Comment by TomasVotruba** [comment]
> In this case, it's just blindly converting the annotations to attributes, but the bug here it shouldn't do that. It should check what attributes already exist before adding new ones.

This would be problematic, as some annotations can be added once, e.g. mentioned Doctrine ID, another multiple times, e.g Symfony routes. That would mean to map every attribute with such a manual marker. Huge amount of complexity, instead of using attributes they way they're common to use.

As stated, the original annotation should be removed, as there is no point of it for PHPUnit 10.

**Comment by dkarlovi** [comment]
Attributes already communicate if they're allowed to be added multiple times. If not, you shouldn't add another one.

In this case, Rector will produce invalid code:

```
Message:  Attribute "PHPUnit\Framework\Attributes\Small" must not be repeated
```

**Comment by TomasVotruba** [comment]
I see :+1:  Could you add a failing test fixture in https://github.com/rectorphp/rector-phpunit for this case?
Maybe it will be more clear what we can do there.

**Comment by dkarlovi** [comment]
I don't think this is a PHPUnit specific issue, it's working with attributes in general:

if you want to add an attribute:
1. check current attributes, is it already there?
2. if so, is it `\Attribute::IS_REPEATABLE`?
3. if not there or repeatable, add the attribute, otherwise don't add it

I can add a fixture to PHPUnit repo, but it will probably need to be solved here?

The `CoversNothing` situation is a IMO simple extension of this pattern where you allow specific attributes to be blocked by other attributes, like so:

if you want to add the `CoversNothing` attribute:
1. check current attributes, is it already there?
2. if so, is it `\Attribute::IS_REPEATABLE`?
4. if so, is the attribute blocked? (new check)
4. if not there or repeatable or blocked, add the attribute, otherwise don't add it

This would mean you'd be able to define a mapping which can block specific attributes from being added, something like

```yaml
blocked:
    CoversNothing:
        - CoversClass
```

**Comment by TomasVotruba** [comment]
It can be added here as well :+1:
